packages/devkit/src/utils/package-json.ts

```diff
@@ -2,17 +2,59 @@ import { readJson, updateJson } from 'nx/src/generators/utils/json';
 import { installPackagesTask } from '../tasks/install-packages-task';
 import type { Tree } from 'nx/src/generators/tree';
 import { GeneratorCallback } from 'nx/src/config/misc-interfaces';
+import { gt } from 'semver';
+
+const NON_SEMVER_TAGS = {
+  next: 1,
+  latest: 0,
+  previous: -1,
+  legacy: -2,
+};
 
 function filterExistingDependencies(
   dependencies: Record<string, string>,
-  existingDependencies: Record<string, string>
+  existingAltDependencies: Record<string, string>
 ) {
-  if (!existingDependencies) {
+  if (!existingAltDependencies) {
     return dependencies;
   }
 
   return Object.keys(dependencies ?? {})
-    .filter((d) => !existingDependencies[d])
+    .filter((d) => !existingAltDependencies[d])
+    .reduce((acc, d) => ({ ...acc, [d]: dependencies[d] }), {});
+}
+
+function updateExistingDependenciesVersion(
+  dependencies: Record<string, string>,
+  existingAltDependencies: Record<string, string>
+) {
+  return Object.keys(existingAltDependencies)
+    .filter((d) => {
+      if (!dependencies[d]) {
+        return false;
+      }
+
+      const incomingVersion = dependencies[d];
+      const existingVersion = existingAltDependencies[d];
+
+      if (
+        incomingVersion in NON_SEMVER_TAGS &&
+        existingVersion in NON_SEMVER_TAGS
+      ) {
+        return (
+          NON_SEMVER_TAGS[incomingVersion] > NON_SEMVER_TAGS[existingVersion]
+        );
+      }
+
+      if (
+        incomingVersion in NON_SEMVER_TAGS ||
+        existingVersion in NON_SEMVER_TAGS
+      ) {
+        return true;
+      }
```

Comment:
This portion is up for debate @FrozenPandaz.
`semver` can't parse npm tags, so it throws an error.

I created a map to create an order of preference for some common npm tags. The order of preference is up for debate, but also the usage of it.

We could ignore having an order of preference and instead simply say, if an npm tag is used, just use the incoming version tag.

What are your thoughts?